feat(linter): support `--print-config` to print config file for project

mysteryven opened the pull request:
Continue work on #4742.

mysteryven commented:
* **#6579** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6579?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @mysteryven and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

pumano commented:
very useful feature for debugging!

mysteryven commented on apps/oxlint/src/lint/mod.rs:
```diff
@@ -116,6 +116,14 @@ impl Runner for LintRunner {
         enable_plugins.apply_overrides(&mut oxlintrc.plugins);
+
+        // TODO: add support for printing the configuration for a specific path when overrides supported.
+        let oxlintrc_for_print = if let Some(PrintConfig::All) = &misc_options.print_config {
```
> even advertise it as a CLI option

I wanted `--print-config`/`--print-config all` to represent 'all', but I didn't find a proper way in bpaf, to avoid breaking change, maybe warn for now?

pacak commented on apps/oxlint/src/command/mod.rs:
```diff
@@ -25,6 +25,31 @@ pub struct MiscOptions {
+    Path(PathBuf),
+}
+
+impl std::str::FromStr for PrintConfig {
```
`PathBuf` can handle non-utf8 inputs since they are valid in Windows/Linux, `FromStr` can handle only valid utf8 so doing this you break a bit of compatibility. In practice this is not a big problem, but I wouldn't do it unless there's a reason.

I'm assuming you want to treat `--print-config` as "let's check all the configs" and `--print-config foo` as "let's check the config in `foo`.

You want two parsers for each case, both parsers should produce `PrintConfig` type. IT is perfectly valid to have multiple parsers with the same name, here - `--print-config`, you want to try to run that parses a file name first. Easiest way would be this is something like this:

```
#[derive(Debug, Clone, Default, Bpaf)]
pub enum PrintConfig {
    Path(
        /// help if you want
        #[bpaf(long("print-config"))] PathBuf),

    /// Print the configuration for all files
    #[default]
    #[bpaf(long("print-config"))]
    All,
    
}
```

DonIsaac commented on apps/oxlint/src/command/mod.rs:
```diff
@@ -25,6 +25,31 @@ pub struct MiscOptions {
+    Path(PathBuf),
+}
+
+impl std::str::FromStr for PrintConfig {
```
Why not `From<&str>` if this always returns `Ok`?

mysteryven replied:
Sorry, it's auto generated by editor, I didn’t notice this.

DonIsaac commented on apps/oxlint/src/command/mod.rs:
```diff
@@ -25,6 +25,31 @@ pub struct MiscOptions {
+}
+
+impl std::str::FromStr for PrintConfig {
+    type Err = String;
```
If you do decide to stick with `FromStr`, consider using `Infallible`

pacak replied:
`bpaf` constraints arbitrary arguments to be a `FromStr`. One way would be to consume `String` or `OsString` and convert it into `PrintConfig` with `map`, second way is to parse into `enum` directly as I described.

DonIsaac commented:
LGMT

DonIsaac commented:
### Merge activity

* **Oct 22, 1:55 PM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

mysteryven committed:
feat(linter): support `--print-config all` to print config file for project (#6579)

Continue work on #4742.

Only `oxlint --print-config all` is supported. It's useful to migrate from command-line interface to config file.

The `--print-config PATH` looks not really useful for us now, I will add it after config file overrides supported.
